Fix: Reponse export fails to include all data

mwithheld opened the pull request:
"Download in text format" fails to include the response data for some question types in versions 3.1.5+ on Oracle DB

mwithheld commented:
Fixes this PHP Notice in the logs:
PHP Notice:  Undefined property: stdClass::$response in mod/questionnaire/questionnaire.class.php on line 3105
1. {main}() mod/questionnaire/report.php:0
2. questionnaire->generate_csv() mod/questionnaire/report.php:484

mwithheld commented:
Seen at url /mod/questionnaire/report.php?instance=2323&action=dwnpg&group=0

mwithheld commented:
This bug was seen and the fix tested on Moodle version 3.1, Questionnaire version 3.1.7.

mchurchward commented:
Can you provide test steps I can use to see the problem and then verify the fix?
Thanks.

mwithheld commented:
- Using a Oracle DB-drive Moodle
- Create a mod_questionnaire
- Add a dropdown question
- Add a  textbox question
- Save
- Answer the questions
- As instructor/admin, export the answers (questionnaire item > View All Responses > Download all in text format)
- When I view the downloaded responses file, it does show the textbox data

Output with added debug lines:
`
[13-Jun-2017 12:37:35] generate_csv::Looking at $questionobj=mod_questionnaire\question\text Object
(
    [id] => 2430
    [name] => Donuts
    [type] => Text Box
    [choices] => Array
        (
        )

    [responsetable] =>
    [length] => 20
    [precise] => 25
    [position] => 9
    [content] => <p>Spell donuts backward five times.</p>
    [allchoices] =>
    [required] => n
    [deleted] => n
    [survey_id] => 2323
    [dependquestion] => 0
    [dependchoice] => 0
    [type_id] => 2
    [response_table] => response_text
    [context] => context_module Object
        (
            [_id:protected] => 1035262
            [_contextlevel:protected] => 70
            [_instanceid:protected] => 709356
            [_path:protected] => /1/430672/989202/1035262
            [_depth:protected] => 4
        )

    [response] => mod_questionnaire\response\text Object
        (
            [question] => mod_questionnaire\question\text Object
 *RECURSION*
        )

)

[13-Jun-2017 12:37:35] generate_csv::Looking at $responserow=stdClass Object
(
    [id] => 2801_essaybox_2963
    [submitted] => 1497379717
    [complete] => y
    [grade] => 0
    [username] => markv
    [firstnamephonetic] =>
    [lastnamephonetic] =>
    [middlename] =>
    [alternatename] =>
    [firstname] => Mark
    [lastname] => van Hoek
    [department] => SYST
    [institution] =>
    [userid] => 22
    [rid] => 2801
    [question_id] => 2430
    [choice_id] => 0
    [dbms_lob.substr(qro.response,1000,1)] => donuts backward donuts ba
    [rank] => 0
)

[13-Jun-2017 12:37:35] PHP Notice:  Undefined property: stdClass::$response in mod/questionnaire/questionnaire.class.php on line 3105
[13-Jun-2017 12:37:35] PHP Stack trace:
[13-Jun-2017 12:37:35] PHP   1. {main}() mod/questionnaire/report.php:0
[13-Jun-2017 12:37:35] PHP   2. questionnaire->generate_csv() mod/questionnaire/report.php:484
`

That
[dbms_lob.substr(qro.response,1000,1)] => 
part should say 
[response]=>

mchurchward commented:
Couple of questions on the test...
1. Does it only cause a problem on Oracle databases? If so, can you indicate why?
2. Your patch changes code in the `classes/response/multiple.php` file, while your test does not include questions that use the multiple response class. Is it the absence of these questions that causes the problem?

mwithheld commented:
Good questions, MIke.

1. I've only tested Oracle.  I think Oracle is just not able to translate "dbms_lob.substr(qro.response,1000,1)" into the actual field name, "response".  Note that in Oracle, the text questionnaire_response_other.response field becomes a CLOB(4000), which is why it gets wrapped by Moodle core into dbms_lob.substr().

2. No, but that's a good question.  Hm...

**Further testing:**
_Hypothesis: Adding a checkbox fixes the problem:_
- Add a checkbox question (response type = multiple) as the last question (order=dropdown, text, checkboxes) -- export has all dropdown data, missing all text field data, includes all checkbox data
- Move the dropdown below the textbox (order=text, dropdown, checkboxes) -- export has all dropdown data, all text field data, includes all checkbox data
- Move the dropdown back up above the textbox (order=dropdown, text, checkboxes) -- export has all dropdown data, missing all text field data, includes all checkbox data
- Hypothesis fails



I'd missed a stacktrace line in the debug output (now added above and reproduced here):
[05-Jul-2017 10:55:30] PHP Notice:  Undefined property: stdClass::$response in /moodle_static/moodle/mod/questionnaire/questionnaire.class.php on line 3103

 -- looking further at my notes I see that we first saw the problem for checkboxes (response type=multiple), and fixed it there in multiple.php, but later also found the problem with dropdowns (response type=single) but I failed to fix it there.


_Hypothesis: The problem exists for checkboxes and dropdowns:_
- Remove the dropdown, move the checkboxes up to be first (order=checkboxes, text) -- export has all checkbox data, missing all text field data
- Hypothesis confirmed

So my suggestion is to go further with my suggested fix -- in each of mod/questionnaire/classes/response/...:
- multiple.php
- single.php
Replace "$DB->sql_order_by_text('qro.response', 1000) . '" with "$DB->sql_order_by_text('qro.response', 1000) . ' AS response".

I noticed that:
- base.php as 'AS'.$field already in there.
- boolean.php already has the "AS response" part.
